api/envoy/config/core/v3/protocol.proto

```diff
@@ -157,6 +157,16 @@ message Http1ProtocolOptions {
   //   - Not a response to a HEAD request.
   //   - The content length header is not present.
   bool enable_trailers = 5;
+
+  // Allows Envoy to process requests/responses with both `Content-Length` and `Transfer-Encoding`
+  // headers set. By default such messages are rejected, but if option is enabled - Envoy will
+  // remove Content-Length header and process message.
+  // See `RFC7230, sec. 3.3.3 <https://tools.ietf.org/html/rfc7230#section-3.3.3>` for details.
+  //
+  // .. attention::
+  //   Enabling this option might lead to request smuggling vulnerability, especially if traffic
+  //   is proxied via multiple layers of proxies.
+  bool allow_chunked_length = 6;
```

Comment:
Would it be appropriate to add either/both of these annotations:
[
    (udpa.annotations.security).configure_for_untrusted_downstream = true,
    (udpa.annotations.security).configure_for_untrusted_upstream = true
  ];

Reply:
I'm not quite sure.
You might want to configure it in case of misbehaving downstream/upstream, but default behavior (reject request/response if both headers are present)  is safe and in general envoy users will not need to change it.
Or am I interpreting these annotations wrong?

Reply:
cc @htuch 

I'm not an expert in this area.  At one point someone convinced me that closing the connection in cases where HTTP messages contained content-length and transfer-encoding: chunked was safer than simply removing the content-length.  I think that handling requests with both content-length and transfer-encoding chunked by removing content-length leaves you vulnerable to HTTP de-synchronization attacks if a vulnerable proxy is sitting in front of the Envoy.  Replying with an error and closing the connection may result in safer behavior, but I wouldn't be surprised if de-synchronization were still possible if a vulnerable proxy is sitting in front of the envoy.

RFC 7230 section 3.3.3, bullet point 3 considers this situation to be an error.  It does seem to allow for terminating the request or forwarding it along after removing the content-length header, and those are two options you allow users to select via this config knob.  The most recent revision from 2020 also has similar wording.  I think we are following best practices, I just don't know what level of warning is most appropriate here.

Reply:
IMO treating it as an error by default is appropriate and we have the right attention/warning on this to tell people about the risk, but I'm not an expert in this area. @PiotrSikora any thoughts here? @htuch?

Reply:
I think that simply removing `Content-Length` would be fine, but I have no objections if @antoniovicente prefers rejecting requests. Both are valid options.

Reply:
From an API perspective, this seems the right choice and we do call attention to the danger of enabling. I'd actually like to understand why this is needed though, is there some real world traffic that needs this? Can we explain this usage in the commit message and/or docs.

Reply:
>is there some real world traffic that needs this? 
In our case - some old API clients behave that way. And as nginx ignores Content-Length by default in this case it was unnoticed until we tried to serve this traffic via Envoy.

Reply:
Seems legit; even if the old API clients are not RFC conformant, I don't see an issue allowing opt-in. You might want to add to https://www.envoyproxy.io/docs/envoy/latest/configuration/best_practices/edge#best-practices-edge a section on "what not to configure" and include this.

Reply:
I think we're doing the right thing by adding the config knob and setting its default to reject.  I think that the remaining question was what to include in the warning and if we wanted to set some security annotations like configure_for_untrusted_downstream on this field.

I think that the comment you added is sufficient.  I'm fine not adding extra security annotations to the config field.  We can always revisit later.

Reply:
Yeah, the annotations are for fields which should be _set_ for untrusted downstream. In this case, we have a field that should not be set for untrusted downstream, so I think it's fine to skip.